packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -680,26 +838,52 @@ abstract class State<T extends StatefulWidget> {
   ///
-  /// The given build context object contains information about the location in
-  /// the tree at which this widget is being built. For example, the context
-  /// provides the set of inherited widgets for this location in the tree.
+  ///  * After calling [initState].
+  ///  * After calling [didUpdateConfig].
+  ///  * After receiving a call to [setState].
+  ///  * After a dependencies of this [State] objects changes (e.g., an
+  ///    [InheritedWidget] referenced by the previous [build] changes).
+  ///  * After calling [deactivate] and then reinserting the [State] object into
+  ///    the tree at another location.
   ///
-  /// The context argument is always the same as [State.context]. This argument
-  /// is provided redundantly here to match the [WidgetBuilder] function
-  /// signature used by [StatelessWidget.build] and other widgets.
+  /// The framework replaces the subtree below this widget with the widget
+  /// returned by this method, either by updating the existing subtree or by
+  /// removing the subtree and inflating a new subtree, depending on whether the
+  /// widget returned by this method [Widget.canUpdate] the root of the
+  /// existing subtree.
+  ///
+  /// Typically implementations return a newly created constellation of widget
+  /// that are configured with information from this widget's constructor, the
+  /// given [BuildContext], and the internal state of this [State] object.
+  ///
+  /// The given [BuildContext] contains information about the location in the
+  /// tree at which this widget is being built. For example, the context
+  /// provides the set of inherited widgets for this location in the tree. The
+  /// [BuildContext] argument is always the same as the [context] property of
+  /// this [State] object and will remain the same for the lifetime of this
+  /// object. The [BuildContext] argument is provided redundantly here so that
+  /// this method matches the signature for a [WidgetBuilder].
   @protected
   Widget build(BuildContext context);
 
   /// Trampoline to make the [build] closure library-accessible.
   WidgetBuilder get _build => build;
 
-  /// Called when an Inherited widget in the ancestor chain has changed. Usually
-  /// there is nothing to do here; whenever this is called, build() is also
-  /// called.
+  /// Called when a dependencies of this [State] objects changes.
+  ///
+  /// For example, if this previous call to [build] referenced an
+  /// [InheritedWidget] that later changed, the framework will call this
+  /// method to notify this object about the change.
+  ///
+  /// Subclasses rarely override this method because the framework always
+  /// calls [build] after a dependency changes. Some subclasses do override
+  /// this method because they need to do some expensive work (e.g., network
+  /// fetches) when their dependencies change, and that work would be too
+  /// expensive to do for every build.
```

Comment:
The only client I found outside of framework.dart was _ImageState, which uses this callback to call _resolveImage, which I think can trigger network traffic.  At least, that's the example I was trying to base this on.